kernel: add MultiPath TCP kernel config & module

sKyissKy opened the pull request:
add MultiPath TCP kernel config & module

Ansuel commented:
Hi we support 5.15 and 6.1 kernel version. You can drop each check for the kernel version. Also please drop the config-kernel option and lets just add the kernel modules.

Also please check the guideline on how to submit a patch with adding to this a correct tag and adding a commit description.

luizluca commented on config/Config-kernel.in:
```diff
@@ -1054,6 +1054,13 @@ if KERNEL_IP_MROUTE
 
+config KERNEL_MPTCP
+	bool "Enable IPv4 MultiPath TCP support"
+	default y
```
default y? How much extra code will it add to the image?

sKyissKy replied:
already fixed.

john-tho commented:
@sKyissKy, you have already been told not to spam pull requests:
https://github.com/openwrt/openwrt/pull/11147#issuecomment-1304643417
https://github.com/openwrt/openwrt/pulls?q=is%3Apr+author%3AsKyissKy+is%3Aclosed
You just waste people's time and compute resources, especially when you are being told exactly the same things: https://github.com/openwrt/openwrt/pull/11147#issuecomment-1453078866. https://github.com/openwrt/openwrt/pull/11064#issuecomment-1288217320
You can use `git rebase` to edit your commits, then `git push --force` to update your PR branch.

If you had addressed the feedback in any of your previous pull requests, this would have been merged a long time ago.

sKyissKy commented:
> Hi we support 5.15 and 6.1 kernel version. You can drop each check for the kernel version. Also please drop the config-kernel option and lets just add the kernel modules.
> 
> Also please check the guideline on how to submit a patch with adding to this a correct tag and adding a commit description.

If set MPTCP, MPTCP_IPV6 as a kernel module, an empty packet will be generated.

john-tho commented:
You need to depend on your Config-kernel.in option in your kernel package.
Otherwise, because `MPTCP` is a Kconfig bool, buildbot would compile this into every kernel via `ALL_KMODS`.

This was detailed months ago on one of your duplicate PRs; https://github.com/openwrt/openwrt/pull/11147#issuecomment-1303170026

luizluca commented:
A PR should not be a log of your changes over time but the logical blocks required to apply your changes. If you need to fix an old commit, you need to squash them together (rebase and squash/fixup). Afterwards, you can push -f your current final work for review. There are tons of instructions on the internet.

The commit titles are not using the correct prefixes used by openwrt. Check the git log and you'll see there is a logic. You can see the tags prefixes by using git log -- aDir/aFile.

You need to describe, at least, what and why in your commit messages. "Update file foo" says nothing. Again, check the existing log to have an idea.

sKyissKy commented:
> A PR should not be a log of your changes over time but the logical blocks required to apply your changes. If you need to fix an old commit, you need to squash them together (rebase and squash/fixup). Afterwards, you can push -f your current final work for review. There are tons of instructions on the internet.
> 
> The commit titles are not using the correct prefixes used by openwrt. Check the git log and you'll see there is a logic. You can see the tags prefixes by using git log -- aDir/aFile.
> 
> You need to describe, at least, what and why in your commit messages. "Update file foo" says nothing. Again, check the existing log to have an idea.

I edit the code directly on the github webpage. It seems that the github webpage version does not provide the function of rebase and merge commit. If the above functions can be realized on the webpage, please advise.

luizluca commented:
> > A PR should not be a log of your changes over time but the logical blocks required to apply your changes. If you need to fix an old commit, you need to squash them together (rebase and squash/fixup). Afterwards, you can push -f your current final work for review. There are tons of instructions on the internet.
> > 
> > The commit titles are not using the correct prefixes used by openwrt. Check the git log and you'll see there is a logic. You can see the tags prefixes by using git log -- aDir/aFile.
> > 
> > You need to describe, at least, what and why in your commit messages. "Update file foo" says nothing. Again, check the existing log to have an idea.
> 
> I edit the code directly on the github webpage. It seems that the github webpage version does not provide the function of rebase and merge commit. If the above functions can be realized on the webpage, please advise.

I don't believe it's possible. GitHub web is very basic. You need to edit it locally.

john-tho commented on package/kernel/linux/modules/netsupport.mk:
```diff
@@ -1434,6 +1434,23 @@ endef
+define KernelPackage/inet-mptcp-diag
+  SUBMENU:=$(NETWORK_SUPPORT_MENU)
+  TITLE:=INET diag support for MultiPath TCP
+  DEPENDS:= +@KERNEL_MPTCP +@KERNEL_MPTCP_IPV6 +kmod-inet-diag
+  KCONFIG:= CONFIG_INET_MPTCP_DIAG@ge5.6
+  FILES:= $(LINUX_DIR)/net/mptcp/mptcp_diag.ko@ge5.6
```
no `+` before the `@CONFIG_SYMBOLS`, `+` is for adding packages.

no need for the `@ge5.6` kernel 5.6 version switches after Kconfig symbol nor in file list (this branch now supports 5.15 and 6.1)

john-tho commented:
Great to see a single commit with an appropriate title now.
You need to add a description to the commit message (explain why this change was made),  and your Signed-off-by tag.

Neustradamus commented:
To follow

Djfe commented:
for what @john-tho asked:
commandline:
`git commit --amend --sign-off`
it edits the last commit message and adds your sign off for you :)

xback commented:
@sKyissKy 
Hi,
Just wondering if you will finish this one? :-)
I'm interested

Thanks!

rodsmar commented:
I've made a PR based on this. Check #14323

sKyissKy committed:
kernel: add MultiPath TCP kernel config & module

Add MultiPath TCP support by setting the config and module files.

Multipath TCP (MPTCP) is an effort to allow simultaneous use of multiple
IP addresses/interfaces through a modification of TCP that presents a
regular TCP interface to applications, while actually spreading data
across multiple subflows.

The benefits of this module include better resource utilization, better
throughput, and smoother occurrence of failures
(font: https://openwrt.org/docs/guide-user/network/mptcp)

Co-authored-by: Rodrigo B. de Sousa Martins <rodrigo.sousa.577@gmail.com>
Signed-off-by: sKy King <29267720+sKyissKy@users.noreply.github.com>
Link: https://github.com/openwrt/openwrt/pull/12884
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

hauke commented on config/Config-kernel.in:
```diff
@@ -1134,6 +1134,12 @@ if KERNEL_IP_MROUTE
 
+config KERNEL_MPTCP
+	bool "Enable IPv4 MultiPath TCP support"
+	default n
```
```suggestion
```
`default n` is not needed, `n` is already the default.

hauke commented on config/Config-kernel.in:
```diff
@@ -1175,6 +1181,13 @@ if KERNEL_IPV6
 
+	config KERNEL_MPTCP_IPV6
+		bool "Enable IPv6 MultiPath TCP support"
+		default n
```
```suggestion
		default y
```
I assume that most users that have selected `KERNEL_MPTCP` also want to have IPv6 support. 
Please make this active by default.

hauke commented on package/kernel/linux/modules/netsupport.mk:
```diff
@@ -1495,6 +1495,23 @@ endef
+define KernelPackage/inet-mptcp-diag
+  SUBMENU:=$(NETWORK_SUPPORT_MENU)
+  TITLE:=INET diag support for MultiPath TCP
+  DEPENDS:= +KERNEL_MPTCP +kmod-inet-diag
```
Please do not add the `KERNEL_MPTCP` configuration here, bgut depend on it:
```suggestion
  DEPENDS:= @KERNEL_MPTCP +kmod-inet-diag
```

hauke commented:
Please add your real name as author and in the `Signed-off-by:` line.

matttbe commented:
Hello,

I guess this PR can be closed now that #16786 has been merged.

hauke commented:
Multipath TCP support was added in c8d5abd62b70137c70bf12e83b6d0708b980abb3 and a2aabc9a7a4ec8a66477fbdde84a73b5942b460e using #16786.
